core/arch/arm/mm/tee_mmu.c

```diff
@@ -74,84 +66,88 @@
 /* Support for 31 concurrent sessions */
 static uint32_t g_asid = 0xffffffff;
 
-static void tee_mmu_umap_set_pa(struct tee_mmap_region *tbl,
-			size_t granule, paddr_t pa, size_t size, uint32_t attr)
-{
-	paddr_t upa = ROUNDDOWN(pa, granule);
-	size_t usz = ROUNDUP(pa - upa + size, granule);
-
-	tbl->pa = upa;
-	tbl->size = usz;
-	tbl->attr = attr;
-}
-
-static TEE_Result tee_mmu_umap_add_param(struct tee_mmu_info *mmu, paddr_t pa,
-			size_t size, uint32_t attr)
+static TEE_Result tee_mmu_umap_add_param(struct tee_mmu_info *mmu,
+					 struct param_mem *mem)
 {
-	struct tee_mmap_region *last_entry = NULL;
+	TEE_Result res;
+	struct tee_ta_region *last_entry = NULL;
 	size_t n;
-	paddr_t npa;
+	uint32_t attr = TEE_MMU_UDATA_ATTR;
 	size_t nsz;
+	size_t noffs;
+
+	if (!mobj_is_paged(mem->mobj)) {
+		uint32_t cattr;
+
+		res = mobj_get_cattr(mem->mobj, &cattr);
+		if (res != TEE_SUCCESS)
+			return res;
+		attr |= cattr << TEE_MATTR_CACHE_SHIFT;
+	}
+
+	if (!mobj_is_secure(mem->mobj))
```

Comment:
Next step after this PR is probably to get rid of `mobj_sec_shm` and let `mobj_paged` handle that kind of memory also. The reason I didn't do that in this PR is that I'm trying to keep the changes a small as possible. The primary purpose of `mobj_is_secure()` is to tell how the memory should be mapped. If the MOBJ is OK to use or not is checked earlier.